pkg/R/RDD.R

```diff
@@ -1323,6 +1323,88 @@ setMethod("zipWithIndex",
            lapplyPartitionsWithIndex(x, partitionFunc)
          })
 
+#' Zip an RDD with another RDD.
+#'
+#' Zips this RDD with another one, returning key-value pairs with the
+#' first element in each RDD second element in each RDD, etc. Assumes
+#' that the two RDDs have the same number of partitions and the same
+#' number of elements in each partition (e.g. one was made through
+#' a map on the other).
+#'
+#' @param x An RDD to be zipped.
+#' @param other Another RDD to be zipped.
+#' @return An RDD zipped from the two RDDs.
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' rdd1 <- parallelize(sc, 0:4)
+#' rdd2 <- parallelize(sc, 1000:1004)
+#' collect(zipRDD(rdd1, rdd2))
+#' # list(list(0, 1000), list(1, 1001), list(2, 1002), list(3, 1003), list(4, 1004))
+#'}
+#' @rdname zipRDD
+#' @aliases zipRDD,RDD
+setMethod("zipRDD",
+          signature(x = "RDD", other = "RDD"),
+          function(x, other) {
+            n1 <- numPartitions(x)
+            n2 <- numPartitions(other)
+            if (n1 != n2) {
+              stop("Can only zip RDDs which have the same number of partitions.")
+            }
+
+            # Check if corresponding partitions of both RDDs have the same
+            # number of elements.
+            getPartitionLengthFunc <- function(part) {
+              list(length(part))
+            }
+            partitionLengths1 <- collect(lapplyPartition(x, getPartitionLengthFunc))
```

Comment:
After calling Spark zip(), for serialized RDDs, the content of each partition is of (Array[Byte], Array[Byte]). And in RRDD, when it is passing a tuple2 into worker, it will break the tuple as two elements and pass them to the worker separately:
{code}
             case (key, value) =>
               writeElem(key)
               writeElem(value)
{code}

In worker, it calls readDeserialize() to deserialized the two tuple elements. readDeserialize() will combine the two lists into one list , that is (K1, K2,...,Kn,V1, V2,...,Vm). The problem is that there is no boundary information between Kn and V1. So in partitionFunc(), we can't know the length of keys, and length of values. We can only assume that keys and values have the same length.